@here/harp-mapview/lib/MapView.ts

```diff
@@ -2133,54 +2136,74 @@ export class MapView extends EventDispatcher {
 
-        try {
-            await dataSource.connect();
+            try {
+                await dataSource.connect();
 
-            const alreadyRemoved = !this.m_tileDataSources.includes(dataSource);
-            if (alreadyRemoved) {
-                return;
-            }
-            dataSource.addEventListener(MapViewEventNames.Update, () => {
-                this.update();
-            });
+                const alreadyRemoved = !this.m_tileDataSources.includes(dataSource);
+                if (alreadyRemoved) {
+                    return;
+                }
+                dataSource.addEventListener(MapViewEventNames.Update, () => {
+                    this.update();
+                });
 
-            const theme = await this.getTheme();
-            dataSource.setLanguages(this.m_languages);
+                const theme = await this.getTheme();
+                dataSource.setLanguages(this.m_languages);
 
-            if (theme !== undefined && theme.styles !== undefined) {
-                await dataSource.setTheme(theme);
-            }
+                if (theme !== undefined && theme.styles !== undefined) {
+                    await dataSource.setTheme(theme);
+                }
 
-            this.m_connectedDataSources.add(dataSource.name);
+                this.m_connectedDataSources.add(dataSource.name);
 
-            this.dispatchEvent({
-                type: MapViewEventNames.DataSourceConnect,
-                dataSourceName: dataSource.name
-            });
+                this.dispatchEvent({
+                    type: MapViewEventNames.DataSourceConnect,
+                    dataSourceName: dataSource.name
+                });
 
-            this.update();
-        } catch (error) {
-            logger.error(`Failed to connect to datasource ${dataSource.name}: ${error.message}`);
+                this.update();
+            } catch (error) {
+                logger.error(
+                    `Failed to connect to datasource ${dataSource.name}: ${error.message}`
+                );
 
-            this.m_failedDataSources.add(dataSource.name);
-            this.dispatchEvent({
-                type: MapViewEventNames.DataSourceConnect,
-                dataSourceName: dataSource.name,
-                error
-            });
-        }
+                this.m_failedDataSources.add(dataSource.name);
+                this.dispatchEvent({
+                    type: MapViewEventNames.DataSourceConnect,
+                    dataSourceName: dataSource.name,
+                    error
+                });
+            }
+            resolve();
+        });
+        this.m_promises.push(dsPromise);
+        return dsPromise;
+    }
+
+    /**
+     * Promise that resolves when the theme and all DataSources added via `addDataSource` are added.
+     *
+     * Only necessary if you need to `await` for the `MapView` to be fully loaded before doing some
+     * next step.
+     */
+    async loadPromise() {
```

Comment:
Not sure we should add this API, whats the use case apart from the dispose method not working correctly?

Reply:
That is the only use case I know of so far.

Reply:
Then I think it might not be worth a new API to maintain, at least make it internal, but as discussed above, I think the whole thing boils down to the BackgroundDAtaSource being artificially async which should not be the case and solved there

Reply:
I don't know how you will solve this, because a constructor can't be async. Do you have any ideas?

Reply:
Good point, it would need to basically call most of the content of MapView.addDatasource without the async unneeded stuff, which does not scale... 

If I understand right, we want to fix the dispose method with this... I still tend to thing that adding another function to call before might not solve the issue, as the dispose method can still be called without awaiting... 

So, if I get this right doing something like:

const mapView = new MapView({theme: sometheme});
mapView.dispose();

somehow crashes, caused by the theme loading and and the background datasource being added...

The theme loading can actually be canceled, the MapViewThemeManager is doing so, if a new theme is added... 
if we expose this to be used internally in MapView.dispose() could this solve the issue?

For the BackgroundDataSource, which part of it actually fails on disposal?